src/transports/tcp.ts

```diff
@@ -0,0 +1,236 @@
+    private sm: StreamManagement;
+    private stanzas: Registry;
+    private parser!: StreamParser;
+
+    private socket?: net.Socket;
+    private tlssocket?: TLSSocket;
+    private isSecure: boolean;
+
+    constructor(client: Agent, sm: StreamManagement, stanzas: Registry) {
+        super({ objectMode: true });
+        this.client = client;
+        this.sm = sm;
+        this.stanzas = stanzas;
+        this.isSecure = false;
+
+        this.on('data', e => {
+            this.client.emit('stream:data', e.stanza, e.kind);
+        });
+
+        this.on('end', () => {
+            if (this.client.transport === this) {
+                this.client.emit('--transport-disconnected');
+            }
+        });
+    }
+
+    public initParser(): void {
+        this.parser = new StreamParser({
+            acceptLanguages: this.config.acceptLanguages,
+            allowComments: false,
+            lang: this.config.lang,
+            registry: this.stanzas,
+            wrappedStream: true
+        });
+    
+        this.parser.on('data', async (e: ParsedData) => {
+            const name = e.kind;
+            const stanzaObj = e.stanza;
+            if (name === 'stream') {
+                if (stanzaObj.action === 'open') {
+                    this.hasStream = true;
+                    this.stream = stanzaObj;
+                    return this.client.emit('session:prebind', this.config.jid);
+                }
+                if (stanzaObj.action === 'close') {
+                    this.client.emit('stream:end');
+                    return this.disconnect();
+                }
+            }
+            if (name === 'features') {
+                if (stanzaObj.tls) {
+                    this.write(`<starttls xmlns='urn:ietf:params:xml:ns:xmpp-tls'/>`);
+                }
+            }
+            if (name === 'tls') {
+                if (stanzaObj.type === 'proceed') {
+                    try {
+                        await this.negotiateTls();
+                    } catch (e) {
+                        this.client.emit('--transport-error', e as Error);
+                    }
+                } else if (stanzaObj.type === 'failure') {
+                    this.client.emit('--transport-error', new Error('TLS negotiation failed'));
+                    this.socket?.destroy();
+                    this.emit('end');
+                }
+            }
+
+            this.push({ kind: e.kind, stanza: e.stanza });
+        });
+    
+        this.parser.on('error', (err: any) => {
+            const streamError = { condition: StreamErrorCondition.InvalidXML };
+            this.client.emit('stream:error', streamError, err);
+            this.write(this.stanzas.export('error', streamError)!.toString());
+            return this.disconnect();
+        });
+    }
+
+    async discoverBindings(server: string): Promise<[TransportConfig]> {
+        return [
+            { url: server + ':5222' } as TransportConfig
```

Comment:
I'm guessing this isn't SRV record aware lol

Reply:
Not yet, this _is_ node-only anyways so SRV lookups aren't a problem at all, it's just that it'll literally never be useful because no server I've seen supports SRV records. It honestly seems like it would work to just return 5222 and 5223 and assume one works, maybe that's the fallback though if SRV lookup fails.

Reply:
```suggestion
            // TODO: The spec-correct way is to use SRV DNS records before defaulting.
            //  However, they're rarely (if ever) used, so we don't support them for now.
            { url: server + ':5222' } as TransportConfig
```